src/lib/mina_net2/bitswap_block.ml

```diff
@@ -0,0 +1,237 @@
+  ; last_block_data_size
+  ; full_link_blocks
+  ; non_max_link_block_count
+  ; max_block_size
+  ; max_links_per_block
+  }
+
+let create_schema_length_prefixed ~max_block_size data_length =
+  create_schema ~max_block_size (data_length + 4)
+
+let blocks_of_data ~max_block_size data =
+  if max_block_size <= 2 + link_size then failwith "Max block size too small" ;
+  let max_data_chunk_size = max_block_size - 2 in
+  let data_length = Bigstring.length data in
+  let schema = create_schema ~max_block_size data_length in
+  let remaining_data = ref data_length in
+  let blocks = Blake2.Table.create () in
+  let link_queue = Queue.create () in
+  let dequeue_chunk chunk_size =
+    assert (!remaining_data >= chunk_size) ;
+    let chunk =
+      Bigstring.sub_shared data
+        ~pos:(!remaining_data - chunk_size)
+        ~len:chunk_size
+    in
+    remaining_data := !remaining_data - chunk_size ;
+    chunk
+  in
+  let dequeue_links num_links =
+    assert (Queue.length link_queue >= num_links) ;
+    let links = ref [] in
+    for _ = 1 to num_links do
+      links := Queue.dequeue_exn link_queue :: !links
+    done ;
+    !links
+  in
+  let create_block links chunk_size =
+    let chunk = dequeue_chunk chunk_size in
+    let num_links = List.length links in
+    let size = 2 + (num_links * link_size) + chunk_size in
+    if num_links > absolute_max_links_per_block || size > max_block_size then
+      failwith "invalid block produced" ;
+    let block = Bigstring.create size in
+    Bigstring.set_uint16_le_exn block ~pos:0 num_links ;
+    List.iteri links ~f:(fun i link ->
+        let link_buf = Bigstring.of_string (Blake2.to_raw_string link) in
+        Bigstring.blit ~src:link_buf ~src_pos:0 ~dst:block
+          ~dst_pos:(2 + (i * link_size))
+          ~len:link_size) ;
+    Bigstring.blit ~src:chunk ~src_pos:0 ~dst:block
+      ~dst_pos:(2 + (num_links * link_size))
+      ~len:chunk_size ;
+    let hash = Blake2.digest_bigstring block in
+    Hashtbl.set blocks ~key:hash ~data:block ;
+    Queue.enqueue link_queue hash
+  in
+  (* create the last block *)
+  create_block [] schema.last_block_data_size ;
+  if schema.total_blocks > 1 then (
+    (* create the data-only blocks *)
+    let num_data_only_blocks =
+      schema.total_blocks - schema.full_link_blocks - 1
+      - if schema.non_max_link_block_count > 0 then 1 else 0
+    in
+    for _ = 1 to num_data_only_blocks do
+      create_block [] max_data_chunk_size
+    done ;
+    (* create the non max link block, if there is one *)
+    if schema.non_max_link_block_count > 0 then (
+      let chunk_size =
+        max_block_size - 2 - (schema.non_max_link_block_count * link_size)
+      in
+      create_block (dequeue_links schema.non_max_link_block_count) chunk_size ;
+      (* create the max link blocks *)
+      let full_link_chunk_size =
+        max_block_size - 2 - (schema.max_links_per_block * link_size)
+      in
+      for _ = 1 to schema.full_link_blocks do
```

Comment:
Are you sure this loop should be under `if schema.non_max_link_block_count > 0` condition?

It isn't in the original code

Reply:
I think you are correct. I should also add an explicit edge case for this in the tests, as my fuzzer is too random to encounter this edge case right now I think.